vyper/semantics/utils.py

```diff
@@ -0,0 +1,16 @@
+from typing import Optional
+
+from vyper import ast as vy_ast
+
+
+def get_folded_value(node: vy_ast.VyperNode) -> Optional[vy_ast.VyperNode]:
```

Comment:
`get_folded_value` should probably tag the node with the result if it doesn't find it, that way it's dynamic programming and we don't revisit nodes in the traversal

Reply:
I have moved the tagging of `folded_value` in a node's metadata into the pre-typecheck visitor pass. `get_folded_value` is now a simple wrapper to either get a literal node's value or the literal value of the `folded_value` metadata.